src/referenced-fields.ts

```diff
@@ -294,3 +322,302 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
+//   safe when the root table isn't selected from more than once and it is not
+//   negated in the LF. Right now we conservatively check for the former but
+//   not the second. The negative forms (e.g. it is forbidden that ...) are
+//   not fully supported anyway.
+// - Removing multiple candidates selecting from the same database table to
+//   avoid visibility issues is too conservative. The correct criteria is to
+//   just remove any that are present in at least 2 disjoint subqueries.
+//   Because in this case the problem is that in those cases there is not a
+//   single place in the query that has visibility inside both disjoint
+//   subqueries and that is a requirement for adding the corrent binds for
+//   narrowing.
+// - We can miss subqueries as we don't recurse into everything.
+// - Ban on aggregates is not necessary but solving it requires additional
+//   complexity.
+// - We only care about and recognize `COUNT` aggregates and `GROUP BY`
+//   statements right now.
+// - We assume the ID column is named "id".
+// - This is a very restricted implementation of narrowing which could be
+//   expanded to cover more situations.
+// - Only `SELECT` queries are supported while, theoretically, `UNION` queries
+//   are also valid.
+//
+// This function modifies `abstractSql` in place.
+export const addAffectedIdsBinds = (
+	abstractSql: AbstractSqlQuery,
+	lfRuleInfo: LfRuleInfo,
+) => {
+	const candidates: BindCandidate[] = [];
+	findBindCandidates(['SelectQuery', ['Where', abstractSql]], candidates);
+
+	// Count the number of references for each database table. Multiple
+	// candidates referencing the same table will be removed to conservatively
+	// avoid issues
+	const seenTableNames: { [key: string]: number } = {};
+	for (const candidate of candidates) {
+		if (candidate.tableName in seenTableNames) {
+			seenTableNames[candidate.tableName] += 1;
+		} else {
+			seenTableNames[candidate.tableName] = 1;
+		}
+	}
+
+	// Currently we only apply this simple narrowing to the root table as there
+	// are safety issues otherwise
+	const rootTableCandidate = candidates.find(
+		(candidate) => candidate.alias === lfRuleInfo.rootAlias,
+	)!;
+
+	// In some cases, due to optimizations, the rule's root table is not
+	// selected from in the actual query. We don't narrow in those cases and
+	// neither when the root table is selected from multiple times. The latter
+	// is probably safe but requires more thinking
+	if (!rootTableCandidate || seenTableNames[rootTableCandidate.tableName] > 1) {
+		return;
+	}
+
+	if (rootTableCandidate.scope[0] === 'UnionQuery') {
+		throw new Error('addAffectedIdsBinds only supports SELECT queries');
+	}
+
+	const selectQueryNode = rootTableCandidate.scope as SelectQueryNode;
+	let whereNode: WhereNode | null = null;
+	for (const statement of selectQueryNode.slice(1)) {
+		if (statement[0] === 'Where') {
+			whereNode = statement;
+
+			break;
+		}
+	}
+	if (!whereNode) {
```

Comment:
It's best to explicitly compare against null when that's the intention as it avoids any potential confusion as to the intent (whether the intent is `== null` or `== null || === 0 || === '' || ...`
```suggestion
	if (whereNode == null) {
```